flytepropeller/pkg/compiler/validators/utils.go

```diff
@@ -170,125 +169,244 @@ func UnionDistinctVariableMaps(m1, m2 map[string]*core.Variable) (map[string]*co
+	literalType *core.LiteralType
 }
 
-// ValidateLiteralType check if the literal type is valid, return error if the literal is invalid.
-func ValidateLiteralType(lt *core.LiteralType) error {
-	if lt == nil {
-		err := fmt.Errorf("got unknown literal type: [%v].\n"+
-			"Suggested solution: Please update all your Flyte deployment images to the latest version and try again", lt)
-		return err
-	}
-	if lt.GetCollectionType() != nil {
-		return ValidateLiteralType(lt.GetCollectionType())
+func (t unionInstanceChecker) isInstance(lit *core.Literal) bool {
+	unionType := t.literalType.GetUnionType()
+
+	if u := lit.GetScalar().GetUnion().GetType(); u != nil {
+		found := false
+		for _, d := range unionType.GetVariants() {
+			if AreTypesCastable(u, d) {
+				found = true
+				break
+			}
+		}
+		return found
 	}
-	if lt.GetMapValueType() != nil {
-		return ValidateLiteralType(lt.GetMapValueType())
+
+	// Matches iff we can unambiguously select a variant
+	foundOne := false
+	for _, x := range unionType.GetVariants() {
+		if IsInstance(lit, x) {
+			if foundOne {
+				return false
+			}
+			foundOne = true
+		}
 	}
 
-	return nil
+	return foundOne
 }
 
-// LiteralTypeForLiteral gets LiteralType for literal, nil if the value of literal is unknown, or type collection/map of
-// type None if the literal is a non-homogeneous type.
-func LiteralTypeForLiteral(l *core.Literal) *core.LiteralType {
-	switch l.GetValue().(type) {
-	case *core.Literal_Scalar:
-		return literalTypeForScalar(l.GetScalar())
-	case *core.Literal_Collection:
-		return &core.LiteralType{
-			Type: &core.LiteralType_CollectionType{
-				CollectionType: literalTypeForLiterals(l.GetCollection().GetLiterals()),
-			},
+func getInstanceChecker(t *core.LiteralType) instanceChecker {
+	switch t.GetType().(type) {
+	case *core.LiteralType_CollectionType:
+		return collectionInstanceChecker{
+			literalType: t,
 		}
-	case *core.Literal_Map:
-		return &core.LiteralType{
-			Type: &core.LiteralType_MapValueType{
-				MapValueType: literalTypeForLiterals(maps.Values(l.GetMap().GetLiterals())),
-			},
+	case *core.LiteralType_MapValueType:
+		return mapInstanceChecker{
+			literalType: t,
+		}
+	case *core.LiteralType_Blob:
+		return blobInstanceChecker{
+			literalType: t,
+		}
+	case *core.LiteralType_Schema:
+		return schemaInstanceChecker{
+			literalType: t,
+		}
+	case *core.LiteralType_UnionType:
+		return unionInstanceChecker{
+			literalType: t,
+		}
+	case *core.LiteralType_StructuredDatasetType:
+		return structuredDatasetInstanceChecker{
+			literalType: t,
+		}
+	default:
+		if isNoneType(t) {
+			return noneInstanceChecker{}
+		}
+
+		return trivialInstanceChecker{
+			literalType: t,
 		}
-	case *core.Literal_OffloadedMetadata:
-		return l.GetOffloadedMetadata().GetInferredType()
 	}
-	return nil
+}
+
+func IsInstance(lit *core.Literal, t *core.LiteralType) bool {
+	instanceChecker := getInstanceChecker(t)
+
+	if lit.GetOffloadedMetadata() != nil {
```

Comment:
let's make this just return true... if it's offloaded, just return true.

If you just return true, it might be a false positive.  That means the offloaded literal might actually be of the incorrect type.  But this should be fine... it'll result in a flytekit error at runtime.  that should be fine.

As of this PR, the call-sites for this new function are:
![image](https://github.com/user-attachments/assets/5a9d4765-f5f3-4319-82b3-565d5c67cc72)

All these should be either irrelevant for offloaded, or just result in a runtime error.

cc @EngHabu

Reply:
talked to enghabu about this - ignore this comment for now, and let's keep this function as you have it.  we'll get rid of the inferred type in the future.